### PR TITLE
Start refactor to support handling all creds client-side during the auth flow

### PR DESCRIPTION
**Pull request by rtannenbaum**
Start refactor to support handling all creds client-side during the auth flow.

Since the API jobs are (by design) currently running code from arbitrary integrations, this refactor will allow all raw creds handling to be moved out of the API to more vetted/secure clients.

Changes in this PR:
- Simplify OAuth redirect flow to be entirely client-side. Instead of having the API prepare the import auth URL after the export auth has completed, now both the export and import URLs are obtained before the auth flow, and the redirects all happen on the client. The export and import URLs obtained from the API are stored in session storage along with the other client state.
- Introduce ReserveWorker and GetReservedWorker endpoints. Polling for a reserved worker is moved from the API to the client using these endpoints.

To do:
- #518: Move creds encryption from the API to the client
- #519: Handle code/token exchange which currently still resides in the API. Can either move this to client, too, or remove the (arbitrary) AuthExtensions in favor of a single standardized (and vetted) library.

Tested a photos migration locally.

#461

### CODE REVIEW

File: portability-api/src/main/java/org/datatransferproject/api/action/transfer/ReserveWorkerAction.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public Class<ReserveWorker> getRequestType() {
+        return ReserveWorker.class;
+    }
+
+    @Override
+    public String handle(ReserveWorker reserveWorker) {
+        String id = reserveWorker.getId();
```
Comment (holachuy) [inline portability-api/src/main/java/org/datatransferproject/api/action/transfer/ReserveWorkerAction.java]: Could this be null? Should we validate?

File: client-rest/src/app/transfer/auth-callback.component.ts [hunk 2]
```diff
@@ -31,26 +27,29 @@ export class AuthCallbackComponent implements OnInit {
 
     ngOnInit() {
         let token = this.getToken();
-
         let transferId = this.progressService.transferId();
+        let importUrl = this.progressService.importUrl();
 
         if (Step.AUTHENTICATE_EXPORT === this.progressService.currentStep()) {
```
[the remark this answers is not in the document]
Comment (rtannenbaum) [reply client-rest/src/app/transfer/auth-callback.component.ts]: Done

File: client-rest/src/app/transfer/initiate-transfer.component.ts [hunk 3]
```diff
@@ -33,15 +34,32 @@ export class InitiateTransferComponent implements OnInit {
     }
 
     initiate() {
-        let start: StartTransfer = {
+        let pollForWorkerKey = Observable.interval(1500)
```
[the remark this answers is not in the document]
Comment (rtannenbaum) [reply client-rest/src/app/transfer/initiate-transfer.component.ts]: Done

File: client-rest/src/app/transfer/initiate-transfer.component.ts [hunk 4]
```diff
@@ -33,15 +34,32 @@ export class InitiateTransferComponent implements OnInit {
...
+          .switchMap(() => this.transferService.getReservedWorker({id: this.progressService.transferId()}))
+          .take(20)
```
[the remark this answers is not in the document]
Comment (rtannenbaum) [reply client-rest/src/app/transfer/initiate-transfer.component.ts]: Added an alert

File: portability-api/src/main/java/org/datatransferproject/api/action/transfer/ReserveWorkerAction.java (hunk 1 above)
Comment (rtannenbaum) [reply portability-api/src/main/java/org/datatransferproject/api/action/transfer/ReserveWorkerAction.java]: added validation (apparently Jackson does not honor the annotations we put in that say the field is required)

File: client-rest/src/app/transfer/initiate-transfer.component.ts [hunk 5]
```diff
@@ -34,18 +34,26 @@ export class InitiateTransferComponent implements OnInit {
...
+                pollAttempt++;
                 if (reservedWorker.publicKey) {
-                  console.log("got worker with key: " + reservedWorker.publicKey);
+                  console.log("got transfer worker with public key: " + reservedWorker.publicKey);
                   pollForWorkerKey.unsubscribe();
                   this.progressService.workerReserved(reservedWorker.publicKey);
                   this.startTransferJob();
                   // TODO encrypt creds with worker public key
+                } else if (pollAttempt == maxPollAttempts) {
+                  alert(`Could not poll worker public key`);
```
Comment (holachuy) [inline client-rest/src/app/transfer/initiate-transfer.component.ts]: Maybe could not poll for worker information or something. We don't necessarily need to expose that we're polling for key information.

File: client-rest/src/app/transfer/initiate-transfer.component.ts [hunk 6]
```diff
@@ -34,18 +34,26 @@ export class InitiateTransferComponent implements OnInit {
...
-          .take(20)
+          .take(maxPollAttempts)
           .subscribe(
             reservedWorker => {
-                console.log("polling for worker");
+                console.log("polling for assigned transfer worker");
+                pollAttempt++;
                 if (reservedWorker.publicKey) {
-                  console.log("got worker with key: " + reservedWorker.publicKey);
+                  console.log("got transfer worker with public key: " + reservedWorker.publicKey);
```
Comment (holachuy) [inline client-rest/src/app/transfer/initiate-transfer.component.ts]: // TODO: Remove debug statement in production

**Pull request merged**